onnxruntime/core/framework/memory_info.h

```diff
@@ -0,0 +1,225 @@
+#pragma once
+#include "core/graph/basic_types.h"
+#include "core/framework/alloc_kind.h"
+#include "core/framework/data_types.h"
+#include "core/framework/execution_plan_base.h"
+#include "core/framework/mem_pattern.h"
+#include "core/framework/tensor.h"
+#include "core/framework/ort_value_name_idx_map.h"
+
+#include <iomanip>
+
+namespace onnxruntime {
+using OrtValueIndex = int;
+using OrtValueName = std::string;
+
+struct MemoryInfoPerTensor {
+  MemoryBlock planned_block;
+  MemoryBlock allocated_block;
+};
+
+struct MemoryInfoMap {
+  using MemoryInfoMapT = std::unordered_map<onnxruntime::OrtValueIndex, onnxruntime::MemoryInfoPerTensor>;
+
+ public:
+  MemoryInfoMap() = default;
+
+  void AddPlannedMemory(const OrtValueIndex& idx, const MemoryBlock& mb) {
+    map_[idx].planned_block = mb;
+  }
+
+  void AddAllocMemory(const OrtValueIndex& idx, MemoryBlock& mb) {
+    map_[idx].allocated_block = mb;
+    if (ptr_offset == 0 || (ptr_offset > mb.offset_ && mb.offset_ != 0)) {
+      ptr_offset = mb.offset_;
+    }
+  }
+
+  const MemoryBlock& GetPlannedMemory(const OrtValueIndex& idx) const {
+    auto itr = map_.find(idx);
+    ORT_ENFORCE(itr != map_.end());
+    return itr->second.planned_block;
+  }
+
+  size_t GetPlannedAddress(const OrtValueIndex& idx) const {
+    auto itr = map_.find(idx);
+    ORT_ENFORCE(itr != map_.end());
+    return itr->second.planned_block.offset_;
+  }
+
+  size_t GetPlannedSize(const OrtValueIndex& idx) const {
+    auto itr = map_.find(idx);
+    ORT_ENFORCE(itr != map_.end());
+    return itr->second.planned_block.size_;
+  }
+  size_t GetAllocAddress(const OrtValueIndex& idx, bool raw = false) const {
+    auto itr = map_.find(idx);
+    ORT_ENFORCE(itr != map_.end());
+    if (raw) {
+      return itr->second.allocated_block.offset_;
+    } else {
+      return itr->second.allocated_block.offset_ - ptr_offset;
+    }
+  }
+
+  size_t GetAllocSize(const OrtValueIndex& idx) const {
+    auto itr = map_.find(idx);
+    ORT_ENFORCE(itr != map_.end());
+    return itr->second.allocated_block.size_;
+  }
+
+  bool Contain(const OrtValueIndex& idx) const {
+    return map_.find(idx) != map_.end();
+  }
+
+  void clear() {
+    map_.clear();
+  }
+
+  MemoryInfoMapT::iterator begin() { return map_.begin(); }
+  MemoryInfoMapT::const_iterator begin() const { return map_.begin(); }
```

Comment:
I think it is fine, since I added the type const, it will return a const_iterator, if the object is const (same as cbegin()). So the same begin() can be overloaded.